{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f40e9c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Descargamos datos y preparamos datasets\n",
    "from tech_analysis.train_test_split import TrainTestSets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa317625",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/tech_analysis/train_test_split.py:24: FutureWarning: YF.download() has changed argument auto_adjust default to True\n",
      "  data = yf.download(tickers=ticker, start=start, end=end, interval=interval)\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "\n",
      "1 Failed download:\n",
      "['BTC-USD']: YFPricesMissingError('possibly delisted; no price data found  (6mo 2023-01-01 -> 2025-06-06) (Yahoo error = \"Invalid input - interval=6mo is not supported. Valid intervals: [1m, 2m, 5m, 15m, 30m, 60m, 90m, 1h, 4h, 1d, 5d, 1wk, 1mo, 3mo]\")')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['1m_train', '1m_test', '5m_train', '5m_test', '15m_train', '15m_test', '30m_train', '30m_test', '1h_train', '1h_test', '90m_train', '90m_test', '4h_train', '4h_test', '1d_train', '1d_test', '5d_train', '5d_test', '1wk_train', '1wk_test', '1mo_train', '1mo_test', '3mo_train', '3mo_test', '6mo_train', '6mo_test'])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tts = TrainTestSets()\n",
    "ticker = 'BTC-USD'\n",
    "intervals = [\"1m\", \"5m\", \"15m\", \"30m\", \"1h\", \"90m\", \"4h\", \"1d\", \"5d\", \"1wk\", \"1mo\", \"3mo\", \"6mo\"]\n",
    "# Este es un diccionario\n",
    "data = tts.interval_train_test_split(ticker, intervals)\n",
    "data.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9ac39155",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/santiago/Documents/myst_projects/002_Technical_Analysis/venv/lib/python3.9/site-packages/backtesting/_plotting.py:55: UserWarning: Jupyter Notebook detected. Setting Bokeh output to notebook. This may not work in Jupyter clients without JavaScript support, such as old IDEs. Reset with `backtesting.set_bokeh_output(notebook=False)`.\n",
      "  warnings.warn('Jupyter Notebook detected. '\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"eb9eca0f-1e5b-4315-ac70-e9233bb71943\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(\"eb9eca0f-1e5b-4315-ac70-e9233bb71943\");\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {display_loaded(error);throw error;\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"eb9eca0f-1e5b-4315-ac70-e9233bb71943\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Importamos estrategias\n",
    "from tech_analysis.strategies import SmaCross"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb49269a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optimizamos\n",
    "from tech_analysis.optimization import SMAOptTechAnalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52f5d751",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2f97f1d2'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_4b9a996f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d9465404'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_63c95b8a'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ce6575ac'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_94c10292'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ebf0e621'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_fd77e93d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_9dbe7e4d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_4b1a0c9c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_822a6617'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_72adde5e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_df8b9bac'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ca459466'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_80af1cb5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e6f7bef8'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d7ac87a9'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d14c778e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_fa840f14'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_77208f07'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_58f00f9b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2dfe85ea'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_4a62fbec'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_34bf9a90'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a5ea2bc0'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a03e16e0'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c900ac2c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_83ff576d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_0078fd7e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_86bde643'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_faa734e3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2ae26697'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_02462522'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8e12a140'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8e611502'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_33d40a17'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_76dcb4e8'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_097c5e04'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e452483c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ecf0375d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_fe6f6a0e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_70f3f6af'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_1db46cd2'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ead20e8f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_040d4b89'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2ed0603e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_0f93deec'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_3b35cdd1'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d13de3d6'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bd1560a5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e9a909f9'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_12e07204'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_7b86a682'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f477a38e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_33ecbe68'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8915959a'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2b5dfac5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_83d19a43'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b1bb4ddd'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_4371f185'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a392798a'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bca51fcf'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_05c0cd28'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ef54b849'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_843f529f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d4ed08c3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2efb6b7c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_17857c11'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_1442cdc4'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b988b0fa'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e60b0ad3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_21e6a000'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_10c2445c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_392d7c98'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_18d67de7'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c108c518'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_74a5732e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c8509d20'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bd4fe0df'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_7370bdda'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_7ba515de'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_0fe69a96'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f3b754d6'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_3f104533'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b0d9a4df'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_517a4bc8'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f738b9fb'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_5c0cac7b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_9fccffc2'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_31fbd471'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e3ad8d48'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_90f6db84'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e0c15580'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_465b8d9c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b86bd734'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c2038fc5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_4c4efa44'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_33087ae3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c8f43ffc'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b686424f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_cdf49e29'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_0f5c3e35'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_fe0d3202'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d79bc25d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_87dc1e85'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_3f041822'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d3067ec3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_5f36f4ee'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bf566f84'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b48d2e9e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2f6cbb61'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b4fce9b9'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_1b83950b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_84b9e18f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8cc73cf3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_5663c2c2'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_966c5310'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c08d2d10'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ba689449'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_074af4c3'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_0152aabd'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f0b39fd8'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_958681b7'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_11533979'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c238121b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f56e4525'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_cb0e6358'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8495b8b5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ce43f852'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_669837fc'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b84638d0'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bd3ddf87'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_83613f45'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_07c7cc10'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ad2c64d0'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_aa7065c1'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_76c44a3b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b9be7e21'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_66ee5645'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_3b34ceaa'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ca978bd9'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_90ae6cb9'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_eee72d29'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b99ed3ba'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_cf0ddac4'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a5f70164'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_6e3110dc'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_95b4659c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_1ceac391'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a3783da6'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d701ea98'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_95bcf666'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_92822113'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_32a183f4'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_edf76a7b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_62d19694'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_02f87ff6'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d86b2e53'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b93bd3b0'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a4f90927'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_306ed3ed'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_5d2656db'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_ae91006f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f3586a6c'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8347cb83'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f3407d08'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_96ef23a5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_9fe3a367'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_8abb16af'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_43f2a2fe'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d47a916a'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_80ce73a2'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_79c2d333'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_877d93e1'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_2ac3f46f'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_17810119'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_052ae030'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_e2b8d31b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_f29cfe2a'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_d0da9cba'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_967c59e4'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_b3169d8b'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_9aa5f7ed'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_a5b51f9d'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_9d8127ad'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bc96e51e'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_6a775774'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_bc273ff5'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_242942a6'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_58403d96'\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/santiago/opt/anaconda3/lib/python3.9/multiprocessing/resource_tracker.py\", line 201, in main\n",
      "    cache[rtype].remove(name)\n",
      "KeyError: '/psm_c7a42eb7'\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "OHLC `data` is empty",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m n2 \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m20\u001b[39m,\u001b[38;5;241m60\u001b[39m)\n\u001b[1;32m      3\u001b[0m optimizer \u001b[38;5;241m=\u001b[39m SMAOptTechAnalysis(SmaCross)\n\u001b[0;32m----> 4\u001b[0m \u001b[43moptimizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msma_strategy_optimization\u001b[49m\u001b[43m(\u001b[49m\u001b[43mall_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn2\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/myst_projects/002_Technical_Analysis/tech_analysis/optimization.py:34\u001b[0m, in \u001b[0;36mSMAOptTechAnalysis.sma_strategy_optimization\u001b[0;34m(self, all_data, n1, n2)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m set_type, df \u001b[38;5;129;01min\u001b[39;00m all_data\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m     33\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtrain\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m set_type:\n\u001b[0;32m---> 34\u001b[0m         train_results\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msma_params_n_tf_optimization\u001b[49m\u001b[43m(\u001b[49m\u001b[43mset_type\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn2\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     35\u001b[0m optimal_tf_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(train_results)\n\u001b[1;32m     36\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m optimal_tf_df\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReturn [\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m'\u001b[39m, ascending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/Documents/myst_projects/002_Technical_Analysis/tech_analysis/optimization.py:12\u001b[0m, in \u001b[0;36mSMAOptTechAnalysis.sma_params_n_tf_optimization\u001b[0;34m(self, set_type, data, n1, n2)\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21msma_params_n_tf_optimization\u001b[39m(\u001b[38;5;28mself\u001b[39m, set_type, data, n1, n2):\n\u001b[0;32m---> 12\u001b[0m     bt \u001b[38;5;241m=\u001b[39m \u001b[43mBacktest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrategy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcash\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m10_000_000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcommission\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.002\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     14\u001b[0m     stats, heatmap \u001b[38;5;241m=\u001b[39m bt\u001b[38;5;241m.\u001b[39moptimize(\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;66;03m# Definimos parametros a optimizar\u001b[39;00m\n\u001b[1;32m     16\u001b[0m     n1\u001b[38;5;241m=\u001b[39mn1,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     19\u001b[0m     maximize\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReturn [\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;66;03m# Podemos usar las metricas de rendimiento de stats \u001b[39;00m\n\u001b[1;32m     20\u001b[0m     return_heatmap\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     22\u001b[0m     best_params \u001b[38;5;241m=\u001b[39m heatmap\u001b[38;5;241m.\u001b[39msort_values(ascending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/Documents/myst_projects/002_Technical_Analysis/venv/lib/python3.9/site-packages/backtesting/backtesting.py:1211\u001b[0m, in \u001b[0;36mBacktest.__init__\u001b[0;34m(self, data, strategy, cash, spread, commission, margin, trade_on_close, hedging, exclusive_orders, finalize_trades)\u001b[0m\n\u001b[1;32m   1208\u001b[0m     data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mVolume\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mnan\n\u001b[1;32m   1210\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(data) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m-> 1211\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOHLC `data` is empty\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   1212\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(data\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mintersection({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOpen\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHigh\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLow\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mClose\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mVolume\u001b[39m\u001b[38;5;124m'\u001b[39m})) \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m5\u001b[39m:\n\u001b[1;32m   1213\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`data` must be a pandas.DataFrame with columns \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1214\u001b[0m                      \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOpen\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHigh\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLow\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mClose\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, and (optionally) \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mVolume\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: OHLC `data` is empty"
     ]
    }
   ],
   "source": [
    "n1 = range(5,20)\n",
    "n2 = range(20,60)\n",
    "optimizer = SMAOptTechAnalysis(SmaCross)\n",
    "optimizer.sma_strategy_optimization(all_data=data, n1=n1, n2=n2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71201e2e",
   "metadata": {},
   "source": [
    "Esta es la mejor estrategia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Probamos la mejor estrategia con el dataset de test\n",
    "from backtesting import Backtest\n",
    "\n",
    "test_best_strategy = Backtest(data['1wk_test'], SmaCross, cash=10_000_000, commission=0.002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats = test_best_strategy.run(n1=17, n2=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec31e320",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2806ffbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_best_strategy.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b54d5b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
